concrete/controllers/single_page/login.php

```diff
@@ -388,6 +391,20 @@ public function logout($token = false)
         }
     }
 
+    /**
+     * @param $token
+     * @return RedirectResponse
+     */
+    public function do_logout($token = false)
+    {
+        if ($this->app->make('token')->validate('do_logout', $token)) {
+            $u = new User();
+            $u->logout();
+            $r = Redirect::to('/');
+            return $r;
+        }
```

Comment:
This method needs a consistent return type, I'd do:

```php
// Get a response factory
$factory = $this->app->make(ResponseFactory::class);

// test the token
if ($token->validate('do_logout', $token)) {
    // Resolve the current logged in user and log them out
    $this->app->make(User::class)->logout();

    // Return a new redirect to the homepage.
    return $factory->redirect('/');
}

// Send an error response with a link to log out, so that if CSRF fails it's still easy to log out
return $factory->error('Invalid token, <a href=...>click here</a> to log out.');
```